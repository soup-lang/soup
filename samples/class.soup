class Animal
    ctor(name)
        this.name: name

    func move(x)
        return this.name + ' moved ' + x + ' meters'

class Bird: Animal
    ctor(name, speed) super(name)
        this.speed: speed

    func move(x)
        if this.speed = 1
            return 'Bird ' + super.move(x)
        return 'Bird ' + this.name + ' flied ' + x * this.speed + ' meters'

func makeClass()
    class Cat: Animal
        ctor(name) super(name)
            0

        func move(x)
            return 'Cat ' + this.name + ''' won't move at all'''

        func meow(t, %%)
            if t <= 0
                return
            setTimeout(%, 10 * t)
            console.log(this.name + ': meow ' + t)
            this.meow(t - 1, %%)

    return Cat

a: Animal('Alche')
console.log(a.move(5))

b: Bird('Rokh', 2)
console.log(b.move(4))

d: Bird('Duck', 1)
console.log(d.move(3))

Neko: makeClass()
c: Neko('Cherry')
console.log(c.name)
console.log(c.move(10))
c.meow(3, %%)
