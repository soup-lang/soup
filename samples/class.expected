Alche moved 5 meters
Bird Rokh flied 8 meters
Bird Duck moved 3 meters
Cherry
Cat Cherry won't move at all
Cherry: meow 3
Cherry: meow 2
Cherry: meow 1
